### PR TITLE
Compare contents not just addresses

### PR DESCRIPTION
**Pull request by nburles**
Fixes the GCC slowdown #25 - now Linux build (GCC 6.1.1) runs my test ~15% faster than Windows (VS 2015 update 3) - aggressive optimisations in both cases.  Previously Linux was over 100 times slower.

### CODE REVIEW

**Comment by mike-lischke** [comment]
Interesting changes. I'm surprised that the == template code in CPPUtils.h doesn't kick in automatically when comparing shared_ptr instances. During my debug sessions I found that working.

File: runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp [hunk 1]
```diff
@@ -164,17 +164,17 @@ bool ATNConfigSet::operator == (const ATNConfigSet &other) {
...
     return false;
 
+  if (fullCtx != other.fullCtx || uniqueAlt != other.uniqueAlt ||
+      conflictingAlts != other.conflictingAlts || hasSemanticContext != other.hasSemanticContext ||
+      dipsIntoOuterContext != other.dipsIntoOuterContext) // includes stack context
+    return false;
+
   for (size_t i = 0; i < configs.size(); i++) {
-    if (configs[i] != other.configs[i]) {
+    if (configs[i] != other.configs[i] && *(configs[i]) != *(other.configs[i]))
```
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp]: If we assume here that `configs[i] != other.configs[i]` only compares the content (the pointers) of the 2 shared_ptr why do you require them to be different? In ANTLR comparisons are often done by identity, not equality, but you require both here. I'd expect here that we only need `*(configs[i]) != *(other.configs[i])` here (however as already commented in the pull request, this should already be done by the == overload in CPPUtils.h).

**Comment by nburles** [comment]
It seems that that is what happens using VS, but for some reason not using GCC.  The ATNConfigSet change is the one that particularly fixes Linux performance, but given that it seems to be required, I felt making all of them explicit would be safer to ensure running in Linux does in fact perform the same way as in Windows.

File: runtime/Cpp/runtime/src/atn/LexerActionExecutor.cpp [hunk 2]
```diff
@@ -115,6 +115,10 @@ bool LexerActionExecutor::operator == (const LexerActionExecutor &obj) const {
   return _hashCode == obj._hashCode && Arrays::equals(_lexerActions, obj._lexerActions);
 }
 
+bool LexerActionExecutor::operator != (const LexerActionExecutor &obj) const {
```
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/atn/LexerActionExecutor.cpp]: Why is that needed? The != operator is implicitly defined as being `!(==)` if not declared explicitly.

File: runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp (hunk 1 above)
Comment (nburles) [inline runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp]: I figured that comparing the pointers was quicker than comparing the actual configs, as such if the pointers are the same it can shortcut the full comparison - and if not, it isn't that much of an overhead.

**Comment by mike-lischke** [comment]
I tested mostly with clang, so I cannot say how GCC behaves, but if you found this not working we should probably remove the == operator in CPPUtils.h altogether?

File: runtime/Cpp/runtime/src/atn/LexerActionExecutor.cpp (hunk 2 above)
Comment (nburles) [inline runtime/Cpp/runtime/src/atn/LexerActionExecutor.cpp]: It wouldn't compile in VS2015.3 without it being explicitly declared.

**Comment by nburles** [comment]
To be honest, I hadn't even seen the operator == in CPPUtils.h... VS wasn't flagging it as being relevant (I do most of my debugging in Windows).

It clearly was being used in VS, however, as the performance was fine... but yeah, it might be safer to remove it given that GCC seems to completely ignore it.

File: runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp (hunk 1 above)
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp]: Ok, I'm almost convinced. Only have one issue: what happens with compilers where the == code in CPPUtils.h works? Wouldn't that cause the content check to run twice? Maybe instead you should use `(configs[i].get() != other.configs[i].get() && *(configs[i]) != *(other.configs[i]))`?

Comment (nburles) [inline runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp]: Fair point, it probably is doubly comparing at the moment.

I'd probably lean towards removing the == override and making sure everything still works, rather than changing these particular cases to use .get()... otherwise there may be cases I've missed that work perfectly in VS and clang, but "mysteriously" fail to work correctly under GCC?  (i.e. I think it'd be easier to ensure correctness across compilers without == in CPPUtils.h).

Comment (mike-lischke) [inline runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp]: I agree.

Comment (nburles) [inline runtime/Cpp/runtime/src/atn/ATNConfigSet.cpp]: I'll do a new one - that way if it takes a while to get everything going again, this can still be in and providing reasonable GCC performance in the meantime... and I'm done for the day, so it'll be tomorrow!
